### ISSUE TITLE
Enable Multisite in Wordpress Bitnami Helm Chart

### ISSUE THREAD

**Issue by boquie** [bitnami/charts#4255]
**Which chart**:
Wordpress 9.9.2

**Is your feature request related to a problem? Please describe.**
I would like to use the Bitnami chart for Wordpress Multisite, but I have been unable to get this to work using the parameters currently available in the chart.  

For Multisite to be enabled in Wordpress, an .htaccess file or a .conf file need to be processed.  This should be possible by changing parameters specified in the chart such as allowOverrideNone, htaccessPersistenceEnabled, and customHTAccessCM.  I have tried various combinations of these parameters (along with other necessary modifications to the wp-config.php), but Apache never pick up the changes to the .htaccess file.  The site will run in multisite mode, but without the .htaccess changes Multisite does not work correctly.

**Describe the solution you'd like**
Ideally a parameter could be passed to the chart to indicate that it is a multisite installation.  An alternative solution would be some detailed instructions on how to set up multisite using existing helm parameters.

**Describe alternatives you've considered**
I've tried to set this up manually but have had no success.

**Comment by rafariossaa** [comment]
Hi,
  Would you mind to share the changes you needed in `.htaccess` and what issues did you find when running the multisite ?
  I would like to understand the needed changes before implementing any fix.

**Comment by boquie** [comment]
Hi, thanks for the response.  The .htaccess method:

1.
First I start the chart with no modifications, then I add the necessary parts for multi-site and manually edit wp-config.php.  Here's the relevant part of the wp-config file (I can add the whole file if needed):
```
define('WP_ALLOW_MULTISITE', true);
define('MULTISITE', true);
define('SUBDOMAIN_INSTALL', false);
```
Note that I'm not using subdomains.

At this point multi-site is half working, but the theme for any additional sites that are created in multisite are not loading.  This should be fixed by the missing .htaccess changes below.

2.
From the [Chart documentation](https://github.com/bitnami/charts/tree/master/bitnami/wordpress/#installing-the-chart), in the 'Disabling .htaccess' section I then set the following parameters:

- allowOverrideNone=false
- htaccessPersistenceEnabled=true

I restart the pod and note that the /bitnami/wordpress/.htaccess file is created during the start up process.  I then manually modify the /bitnami/wordpress/.htaccess file and add the following as required for multisite:

```
RewriteEngine On
RewriteBase /
RewriteRule ^index\.php$ - [L]

# add a trailing slash to /wp-admin
RewriteRule ^([_0-9a-zA-Z-]+/)?wp-admin$ $1wp-admin/ [R=301,L]

RewriteCond %{REQUEST_FILENAME} -f [OR]
RewriteCond %{REQUEST_FILENAME} -d
RewriteRule ^ - [L]
RewriteRule ^([_0-9a-zA-Z-]+/)?(wp-(content|admin|includes).*) $2 [L]
RewriteRule ^([_0-9a-zA-Z-]+/)?(.*\.php)$ $2 [L]
RewriteRule . index.php [L]
```

I then restart the pod again.

3.
At this point I'd expect the .htaccess to be fully functional.  My test is to visit:
http://www.my-site.com/wp-admin

Note the missing trailing slash '/'.  If the .htaccess file is functioning, it should add the trailing slash and the admin screen should load, but it does not.

It appears that the apache .conf files are not being updated to allow this to work, but I'm not sure if this is the case. 

Any advise is appreciated, I've also tried using config maps with customHTAccessCM, but I have had a similar result.  I'm fine to use either method you suggest.

**Comment by pestebogdan** [comment]
The way I got it to work involved creating an out-of-helm-chart configmap, and then mounted that using the existing `extraVolumeMounts` and `extraVolumes` variables in **values.yaml** to mount the configmap with the necessary init scripts under **/docker-entrypoint-init.d/** , where the bitnami/wordpress container looks for post-init scripts: https://github.com/bitnami/bitnami-docker-wordpress/tree/master/5/debian-10/rootfs/post-init.d

So something like this worked for me:
```
values.yaml:
...
extraVolumeMounts:
 - name: wordpress-custominit-configmap
   mountPath: /docker-entrypoint-init.d/
extraVolumes:
 - name: wordpress-custominit-configmap
   configMap:
     name: wordpress-custominit-configmap
     defaultMode: 0777
allowOverrideNone: false
htaccessPersistenceEnabled: true
```

..and the following manually-created configmap in the same namespace:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: wordpress-custominit-configmap
data:
  multisite.sh: |
   #!/bin/bash
   chmod +w /bitnami/wordpress/wp-config.php
   wp core install --url=localhost --title=Example --admin_user=dummyusernotusedanyway --admin_password=dummypasswordnotusedanyway --admin_email=user@example.com
   wp core multisite-install --url=localhost --title="Welcome to the WordPress" --admin_user="dummyusernotusedanyway" --admin_password="dummypasswordnotusedanyway" --admin_email="user@example.com"
   sed -i '0,/127.0.0.1/s//localhost/' /bitnami/wordpress/wp-config.php
   wp db query < /docker-entrypoint-init.d/update_db.query
   cat /docker-entrypoint-init.d/.htaccess > /bitnami/wordpress/.htaccess
   chmod -w bitnami/wordpress/wp-config.php
  update_db.query: |
   USE bitnami_wordpress;
   UPDATE `bitnami_wordpress`.`wp_blogs` SET `domain` = 'localhost' WHERE (`blog_id` = '1');
   UPDATE `bitnami_wordpress`.`wp_site` SET `domain` = 'localhost' WHERE (`id` = '1');
  .htaccess: |
   RewriteEngine On
   RewriteBase /
   RewriteRule ^index\.php$ - [L]

   # add a trailing slash to /wp-admin
   RewriteRule ^([_0-9a-zA-Z-]+/)?wp-admin$ $1wp-admin/ [R=301,L]

   RewriteCond %{REQUEST_FILENAME} -f [OR]
   RewriteCond %{REQUEST_FILENAME} -d
   RewriteRule ^ - [L]
   RewriteRule ^([_0-9a-zA-Z-]+/)?(wp-(content|admin|includes).*) $2 [L]
   RewriteRule ^([_0-9a-zA-Z-]+/)?(.*\.php)$ $2 [L]
   RewriteRule . index.php [L]
```

I did it on my local k8s kind cluster so i used localhost as the URL, but you can swap for whatever DNS Name you want. Not really sure why the --url param doesn't work as expected (always puts 127.0.0.1 so that's where i get redirected), that's why i used **wp db query** part of the script to manually replace the URL in the database, and **sed** to modify the entry in wp-config.php.

I just opened a MR today that would bring the configmap part into the helm chart, so if that ends up being approved we will be able to do this with only modifying values in the helm chart: https://github.com/bitnami/charts/pull/4373

**Comment by marcosbc** [comment]
Thanks for sharing @pestebogdan!

In order to support Multisite for the chart in an official way, the container would first need to support such feature.

In the coming months we expect to re-implement the WordPress container image to be based on Bash scripts in the `rootfs` folder (i.e. like [Drupal](https://github.com/bitnami/bitnami-docker-drupal/tree/master/9/debian-10/rootfs)), and as part of those changes we'll also work on adding support for Multisite. However, I'm unable to give you an estimation for when exactly that will happen, so we'll add any updates in this issue.

**Comment by pestebogdan** [comment]
Thanks @marcosbc ; just one last quick update, since the PR #4373 was recently merged, that means the out-of-band ConfigMap is no longer necessary, the Helm chart will create one for you if the value of **customPostInitScripts** is not empty. So now instead of the ConfigMap, something like this using only values.yaml is enough:

```
values.yaml:
...
extraVolumeMounts:
 - name: wordpress-postinit
   mountPath: /docker-entrypoint-init.d/

extraVolumes:
 - name: wordpress-postinit
   configMap:
     name: wordpress-postinit
     defaultMode: 0777

customPostInitScripts:
 multisite.sh: |
  #!/bin/bash
  chmod +w /bitnami/wordpress/wp-config.php
  wp core multisite-install --title="Welcome to the WordPress" --admin_user="dummyuser" --admin_password="dummypassword" --admin_email="user@example.com"
  sed -i '0,/127.0.0.1/s//localhost/' /bitnami/wordpress/wp-config.php
  wp db query < /docker-entrypoint-init.d/update_db.query
  cat /docker-entrypoint-init.d/.htaccess > /bitnami/wordpress/.htaccess
  chmod -w bitnami/wordpress/wp-config.php
 update_db.query: |
  USE bitnami_wordpress;
  UPDATE `bitnami_wordpress`.`wp_blogs` SET `domain` = 'localhost' WHERE (`blog_id` = '1');
  UPDATE `bitnami_wordpress`.`wp_site` SET `domain` = 'localhost' WHERE (`id` = '1');
 .htaccess: |
  <IfModule mod_rewrite.c>
  RewriteEngine On
  RewriteBase /
  RewriteRule ^index\.php$ - [L]

  # add a trailing slash to /wp-admin
  RewriteRule ^([_0-9a-zA-Z-]+/)?wp-admin$ $1wp-admin/ [R=301,L]

  RewriteCond %{REQUEST_FILENAME} -f [OR]
  RewriteCond %{REQUEST_FILENAME} -d
  RewriteRule ^ - [L]
  RewriteRule ^([_0-9a-zA-Z-]+/)?(wp-(content|admin|includes).*) $2 [L]
  RewriteRule ^([_0-9a-zA-Z-]+/)?(.*\.php)$ $2 [L]
  RewriteRule . index.php [L]
  </IfModule>
```

I guess this can be optimised to use some variables instead of hardcoded references to DB name (I used bitnami_wordpress) and Domain (localhost in my example), so feel free to do that if you want, but the above is what I used and works for me.

**Comment by ayuryogalifein** [comment]
I am facing issue while running the above config in AWS. Please guide

MountVolume.SetUp failed for volume "wordpress-postinit" : configmap "wordpress-postinit" not found

**Comment by pestebogdan** [comment]
> I am facing issue while running the above config in AWS. Please guide
> 
> MountVolume.SetUp failed for volume "wordpress-postinit" : configmap "wordpress-postinit" not found
> 
> It seems the configmap is not getting created. Do I need do something else.
> 
> I have also tried creating config map outside of helm. However no luck. Please suggest the needful
> 
> Do I need to change the localhost value in this line sed -i '0,/127.0.0.1/s//localhost/'

Hi @ayuryogalifein ; be sure to match the name of the configmap (generated dynamically using the pattern "<your-release-name>-postinit") with what you put in the "extraVolumes" and "extraVolumeMounts". Here is the configmap template, it might make more sense: https://github.com/bitnami/charts/blob/master/bitnami/wordpress/templates/configmap.yaml

So, in the example i have above with the configmap being named wordpress-postinit, my helm release was named "wordpress". So you you name your helm release something else (ex: ayuryogalifein-wordpress) then the configmap (and what you put under extraVolumes and extraVolumeMounts) would be "ayuryogalifein-wordpress-postinit".

Regarding the sed, instead of localhost you would put the domain name with which you will access wordpress (ex: example.com )

Hope that helps.

**Comment by ayuryogalifein** [comment]
@pestebogdan  Thanks a lot for your suggestion. However, I am getting "Readiness probe warning"
kubectl logs shows 
nami    INFO  wordpress successfully initialized
INFO  ==> Starting gosu...
INFO  ==> Executing /docker-entrypoint-init.d/..2020_12_15_17_32_55.094545708/multisite.sh
Single site database tables already present.
Set up multisite database tables.
Added multisite constants to 'wp-config.php'.
Success: Network installed. Don't forget to set up rewrite rules (and a .htaccess file, if using Apache).

any suggestion?

**Comment by pestebogdan** [comment]
@ ayuryogalifein: Looking at the output it seems to have launched that post-init script. The readiness seems to be doing an httpGet to /wp-login.php so not sure why that is failing for you. I would suggest first looking at the database tables and see if the values that the post-init script modifies are actually there and have the expected values. Then, not sure if you are able to exec into the pod and get a look at /bitnami/wordpress/wp-config.php and see that structure is as you would expect for multisite and .htaccess as well (perhaps temporarily removing the rediness probe while doing the troubleshoot will keep the pod up).

**Comment by ayuryogalifein** [comment]
@pestebogdan thanks for your immediate response. The script seems to be ran successfully. Please find below the result. However, still showing readiness prove and the site is not accessible . Just to add, I am also using the bitnami cert- manager and nginx ingress. Without the multisite config, the site comes up without any issue on https. Am I missing something here?

select * from wp_blogs;                                                                                                 +---------+---------+---------------+------+---------------------+---------------------+--------+----------+--------+------+---------+---------+
| blog_id | site_id | domain        | path | registered          | last_updated        | public | archived | mature | spam | deleted | lang_id |
+---------+---------+---------------+------+---------------------+---------------------+--------+----------+--------+------+---------+---------+
|       1 |       1 | ayuryoga.life | /    | 2020-12-15 19:28:01 | 0000-00-00 00:00:00 |      1 |        0 |      0 |    0 |       0 |       0 |
+---------+---------+---------------+------+---------------------+---------------------+--------+----------+--------+------+---------+---------+


select * from wp_site;                                                                                                  +----+---------------+------+
| id | domain        | path |
+----+---------------+------+
|  1 | ayuryoga.life | /    |
+----+---------------+------+
1 row in set (0.00 sec)



define( 'WP_ALLOW_MULTISITE', true );
define( 'MULTISITE', true );
define( 'SUBDOMAIN_INSTALL', false );
$base = '/';
define( 'DOMAIN_CURRENT_SITE', 'ayuryoga.life' );
define( 'PATH_CURRENT_SITE', '/' );
define( 'SITE_ID_CURRENT_SITE', 1 );
define( 'BLOG_ID_CURRENT_SITE', 1 );

cat /bitnami/wordpress/.htaccess
<IfModule mod_rewrite.c>
 RewriteEngine On
 RewriteBase /
 RewriteRule ^index\.php$ - [L]

 # add a trailing slash to /wp-admin
 RewriteRule ^([_0-9a-zA-Z-]+/)?wp-admin$ $1wp-admin/ [R=301,L]

 RewriteCond %{REQUEST_FILENAME} -f [OR]
 RewriteCond %{REQUEST_FILENAME} -d
 RewriteRule ^ - [L]
 RewriteRule ^([_0-9a-zA-Z-]+/)?(wp-(content|admin|includes).*) $2 [L]
 RewriteRule ^([_0-9a-zA-Z-]+/)?(.*\.php)$ $2 [L]
 RewriteRule . index.php [L]
 </IfModule>

Seeing /wp-login.php HTTP/1.1" 302 in container logs

**Comment by pestebogdan** [comment]
Not sure if it's a copy-paste issue, but the .htaccess you posted seems to have this line wrong: 
`RewriteRule ^([_0-9a-zA-Z-]+/)?(..php)$ $2 [L]`
Should be something like:
`RewriteRule ^([_0-9a-zA-Z-]+/)?(.*\.php)$ $2 [L]`

**Comment by rafariossaa** [comment]
Hi,
  @pestebogdan Thanks you very much for helping with this issue.
  @ayuryogalifein could you try what @pestebogdan indicated ? I also think that this could be the issue.

**Comment by boquie** [comment]
Hi @pestebogdan, thanks for the work on this issue.  I have applied your configuration and it is working almost perfectly.  The only issue I'm having is it doesn't seem to be respecting the .htaccess file.  In particular, if I try to access the admin for the site, **https://www.myurl.com/wp-admin** it doesn't work and I have to add the trailing slash: wp-admin/.  This should be dealt with by .htaccess:

```
# add a trailing slash to /wp-admin
RewriteRule ^([_0-9a-zA-Z-]+/)?wp-admin$ $1wp-admin/ [R=301,L]
```
Here's my working values.yml
```
---
  extraVolumeMounts:
   - name: wordpress-postinit
     mountPath: /docker-entrypoint-init.d/

  extraVolumes:
   - name: wordpress-postinit
     configMap:
       name: wordpress-postinit
       defaultMode: 0777

  allowOverrideNone: false
  htaccessPersistenceEnabled: true

  customPostInitScripts:
   multisite.sh: |
    #!/bin/bash
    chmod +w /bitnami/wordpress/wp-config.php
    wp core multisite-install --title="Welcome to the WordPress" --admin_user="dummyuser" --admin_password="dummypassword" --admin_email="user@example.com"
    sed -i 's/127.0.0.1/www.myurl.com/g' /bitnami/wordpress/wp-config.php
    wp db query < /docker-entrypoint-init.d/update_db.query
    cat /docker-entrypoint-init.d/.htaccess > /bitnami/wordpress/.htaccess
    chmod -w bitnami/wordpress/wp-config.php
   update_db.query: |
    USE bitnami_wordpress;
    UPDATE `bitnami_wordpress`.`wp_blogs` SET `domain` = 'www.myurl.com' WHERE (`blog_id` = '1');
    UPDATE `bitnami_wordpress`.`wp_site` SET `domain` = 'www.myurl.com' WHERE (`id` = '1');
    UPDATE `bitnami_wordpress`.`wp_options` SET `option_value` = 'https://www.myurl.com' WHERE (`option_name` = 'home');
    UPDATE `bitnami_wordpress`.`wp_options` SET `option_value` = 'https://www.myurl.com' WHERE (`option_name` = 'siteurl');
   .htaccess: |
    <IfModule mod_rewrite.c>
    RewriteEngine On
    RewriteBase /
    RewriteRule ^index\.php$ - [L]

    # add a trailing slash to /wp-admin
    RewriteRule ^([_0-9a-zA-Z-]+/)?wp-admin$ $1wp-admin/ [R=301,L]

    RewriteCond %{REQUEST_FILENAME} -f [OR]
    RewriteCond %{REQUEST_FILENAME} -d
    RewriteRule ^ - [L]
    RewriteRule ^([_0-9a-zA-Z-]+/)?(wp-(content|admin|includes).*) $2 [L]
    RewriteRule ^([_0-9a-zA-Z-]+/)?(.*\.php)$ $2 [L]
    RewriteRule . index.php [L]
    </IfModule>

  ingress: 
    certManager: true
    enabled: true
    hostname: www.myurl.com
    tls: true
    
  mariadb: 
    auth: 
      rootPassword: "dummypassword"
      password: "dummypassword"
  wordpressPassword: "dummypassword"
```
Can you confirm this is working for you?
Thanks!

---
Side notes to anyone else working on this: 
- I'm using a Bitnami Wordpress App in Rancher
- I set `option_name` = 'home' and `option_name` = 'siteurl', otherwise page links were broken

**Comment by pestebogdan** [comment]
Hi @boquie. Didn't bring up an environment to validate, but my guess is the reason is that you set `htaccessPersistenceEnabled: True`.

From the description in values.yml (https://github.com/bitnami/charts/blob/c7291dbc6ecc3eed53cfb812aef63924cabb609d/bitnami/wordpress/values.yaml#L131-L135)

```
## Persist the custom changes of the htaccess. It depends on the value of
## `.Values.allowOverrideNone`, when `yes` it will persist `/opt/bitnami/wordpress/wordpress-htaccess.conf`
## if `no` it will persist `/opt/bitnami/wordpress/.htaccess`
##
htaccessPersistenceEnabled: false
```

Via the script you put the desired .htaccess configuration under /bitnami/wordpress/.htaccess , but you told the deployment you would like it to load it from opt/bitnami/wordpress/wordpress-htaccess.conf.

Bogdan.

**Comment by boquie** [comment]
Hi @pestebogdan, thank you for the response, I tried the changing the variable but unfortunately it didn't work.  I've tried several combinations of these variables with no success.  From what I could understand from the [Bitnami docs on .htaccess for Wordpress](https://github.com/bitnami/charts/tree/master/bitnami/wordpress/#disabling-htaccess), this is what should work:

> Also, some plugins permit editing the .htaccess and it might be needed to persit it in order to keep the changes, A way to make it work would be to set htaccessPersistenceEnabled.
> 
> allowOverrideNone=false
> htaccessPersistenceEnabled=true

Any further suggestions would be greatly appreciated.

As an alternative, I could probably try using, wordpress-htaccess.conf, and set it up it as you demonstrated with the standard .htaccess above, but I'll wait and see if there are any other suggestions to get this working.  I would prefer to use the traditional .htaccess since I'm migrating a large existing site and I'm concerned about introducing errors.

**Comment by nknov** [comment]
Has a solution been found, looking for the same

**Comment by marcosbc** [comment]
Hi @nknov, as part of another internal task, we are currently working on adding support for Wordpress Multisite in container environments. Unfortunately I cannot give an ETA, but it should be released in the next 2-4 weeks (it will still take some time because it is a rewrite of the entire container logic). I'll add a note to ping here once it is released.

**Comment by marcosbc** [comment]
Hi all, we just released [version 11.0.0](https://github.com/bitnami/charts/tree/master/bitnami/wordpress#1100) of this chart which includes [Multisite support. Please check [Multisite configuration parameters](https://github.com/bitnami/charts/tree/master/bitnami/wordpress#wordpress-multisite-configuration-parameters) for more info.
